EWL-3549 Style Guide | Refactor Primary Nav

chelsiejohnston opened the pull request:
## JIRA Ticket(s)
- [EWL-3549: Style Guide | Revert Primary Nav changes](https://issues.ama-assn.org/browse/EWL-3549)
- [EWL-3550: Style Guide | Primary Nav review and DRY-ify](https://issues.ama-assn.org/browse/EWL-3550)
- [EWL-2994: Research screenreader tabbing without using multiple H1](https://issues.ama-assn.org/browse/EWL-2994)


## Description:
- Revert the work done in https://github.com/AmericanMedicalAssociation/AMA-style-guide/pull/106 since we decided not to pursue merging in the change to primary nav interactions in Corp. This means that this PR should recreate the current behavior that's on Corp (click to interact with menu).
- Rename most elements in this pattern to be more concise/descriptive per BEM standards.
- Change all the top-level links to `<a>` elements so that one can tab through the main nav and subnavs (this was already done here, but it should hopefully satisfy [EWL-2994](https://issues.ama-assn.org/browse/EWL-2994)
- Refactor (DRY-ify) the accompanying JS.

I hope this JS is better. I think it's better, but I'm not sure anymore. 

I broke out most things (opening/closing menus, setting the active nav item, etc.) into their own functions. The functions are called conditionally based on which element was clicked.


## To Test:
- [ ] Navigate to Templates > Home > Home (or any pattern including the Primary Nav)
- [ ] On mobile, test that the menu works as expected. Compare with Corp.
- [ ] On desktop, test that the menu works as expected. Compare with Corp.


## Automated Test:
N/A (CSS)


## Relevant Screenshots/GIFs:
![3550-a](https://cloud.githubusercontent.com/assets/12160398/26746374/b85d5b22-47b4-11e7-9157-cc32fa91c642.gif)
![3550-b](https://cloud.githubusercontent.com/assets/12160398/26746373/b85d3732-47b4-11e7-9a09-e84769a09b44.gif)


## Remaining Tasks:
- Still need to add the menu's exit animation per https://issues.ama-assn.org/browse/EWL-3108.


## Additional Notes:
My brain hurts


---

[Pull Request Guidelines](https://github.com/palantirnet/development_documentation/blob/master/guidelines/pr_review.md)

chelsiejohnston commented:
@froboy @mrspete this is finally ready for review.

@rpama @deadarm if you want to review this, I'd super appreciate it. I got a little overzealous with renaming/reorganizing here, so LMK if there's anything iffy.

rpama commented on styleguide/source/_patterns/01-molecules/05-navigation/00-primary-nav/primary-nav.twig:
```diff
@@ -1,134 +1,124 @@
-      </div>
+  <ul class="nav-primary_list nav-primary_list-mobile-closed" role="menubar">
+    <li class="nav-primary_section">
+        {% include '04-link-primary-nav.twig' with { 'content': nav.first } %}
```
If you pass a url in to the `04-link-primary-nav.twig` instead of just the dummy `#` I'm seeing the anchor behavior take priority and navigate the user away instead of expanding the menu. We may need some `e.preventDefault()` methods in the js.

chelsiejohnston replied:
This should be accounted for now.

rpama replied:
I am still seeing this issue on desktop. Not sure right now why it's happening, but if you're burned out on this and want me to take a closer look, I will.

![primary_nav_llinks_linking](https://user-images.githubusercontent.com/28711067/26898835-d97872b0-4b92-11e7-8b12-90661c5ef484.gif)

chelsiejohnston commented:
1. This is fixed

2. I _think_ this might be because the `<body>` element is only as tall as the menu button when you're just viewing the molecule. I'm not 100% sure why this would be the case. Maybe some browsers (Chrome?) can handle this and still register clicks to what "should" be the body? I'm a little stumped on this one.

chelsiejohnston commented on styleguide/source/assets/js/nav-primary.js:
```diff
@@ -34,99 +34,99 @@ jQuery.noConflict();
+  // Handling for click events. When someone clicks the nav, the mobile nav button, or anywhere
+  // on the page if the menu is already open:
+
+  $(document).on('click', '.body-nav-primary-open, .nav-primary_section, .nav-primary_button', function(e) {
```
2. This was supposed to detect clicks outside the nav in order to close it, and binding it here/delegating the events seemed like it was necessary to make that work. I agree that it isn't ideal, but I'm not sure about a better solution. Open to try out any ideas you might have, @rpama.

rpama replied:
It turns out my bullet 2 above regarding lost functionality answers my bullet 1 here about touch events. The answer is, "yes, you need a `touchstart` event here." I believe touch devices will treat clicking on things like anchors as click events as weel as touch events, so click listeners capture them, but in this case just touching anywhere on the `.body-nav-primary-open` class doesn't seem to trigger the event.

chelsiejohnston commented:
@rpama I took a stab at addressing the feedback - need to walk away from this for some amount of time. If you have specific ideas about anything, I'm open to looking into those with you.

Otherwise, I'd feel best about letting @deadarm and/or @froboy take a look at this before we merge it in.

rpama commented:
@chelsiejohnston It's looking really close now! I made notes above, one about where we need a `touchstart` listener which should be a pretty straightforward add, and the other still seems to be a bug when the anchors have actual links. If you want my help going through that last one or want to hand it off because you're sick of looking at it, totally pass it on to someone else. This menu is annoying :(

chelsiejohnston commented:
@rpama if you have the availability, go for it! I definitely wouldn't mind. :)

rpama commented:
As per convo with @chelsiejohnston, I'll do a little work on these last few issues and worry about syncing up with [Ewl 3082 store js with components](https://github.com/AmericanMedicalAssociation/AMA-style-guide/pull/118) if we merge that in.

rpama commented:
I went in and added the `touchstart` listener, cleaned up some mixing in the js file between code for this primary-nav and code for the `primary-header` component.

I'm not sure if the issue of passing a url in to the `04-link-primary-nav.twig` instead of a dummy `#` is resolved, but I'm also not 100% certain that it's not a strange artifact of pattern lab. I recommend a quick review of my changes and then we merge this. If there's a genuine issue with those links once we start putting this into the Corp site, we'll make a bug ticket and resolve then.

Could you do a quick review @chelsiejohnston and see my changes look ok still? If so I'll merge it.
